### DISCUSSION TITLE
How to filter Finances.list_financial_events by Marketplace? It returns too many pages in this function.

### DISCUSSION THREAD

**Discussion by paulabirocchi** [saleweaver/python-amazon-sp-api#293] [Q&A] [answered]
I would like to filter the Finances using the function list_financial_events for a specific Market Place:
I'm using the following script:
```
from sp_api.api import `Finances`
from  sp_api.base.marketplaces import Marketplaces

credentials=dict(
        refresh_token=refresh2,
        lwa_app_id=clientid2,
        lwa_client_secret=clientsecret2,
        aws_secret_key=aws_secret_key, # from assume role (secretaccesskey) 
        aws_access_key=aws_access_key, # from assume role (accesskeyid)
        role_arn='arn:aws:iam::XXXXXXX:role/XXXXX' )# from application
```
Then I define the marketplace and call the Finances API, with the marketplace for Netherlands:
```
marketplaces = ['AAAAA'] # Netherlands

finances_obj = Finances(marketplace= Marketplaces.NL,credentials=credentials)
```
Then I create a loop to see how many pages I have in my API call. 
```
nexts = []
for passodetempo in np.arange(0,1000):
	if passodetempo==0: 
		fin2  = finances_obj.list_financial_events(PostedAfter='2021-05-10',PostedBefore='2021-05-11',MaxResultsPerPage=100,MarketplaceIds=','.join(marketplaces))
		next_token = fin2.next_token
		print('passodetempo: '+ str(passodetempo))
		nexts.append(next_token)
	else: 
		fin1  = finances_obj.list_financial_events(PostedAfter='2021-05-10',PostedBefore='2021-05-11',MaxResultsPerPage=100, NextToken=next_token, MarketplaceIds=','.join(marketplaces))
		next_token = fin1.next_token
		print('passodetempo: '+ str(passodetempo))
		nexts.append(next_token)
```
When I look to the results in fin2.payload or fin1.payload the MarketPlaces are not only for Netherlands, but includes also for Germany, France, etc. It doesn't look like this filter for marketplace is working properly. 
In addition, this API call is generating many and many pages, which is suspect as we don't really have some many Orders for NL or with other European countries summed.

Any tip is helpful. Thank you.

**Comment by saleweaver** [comment, marked as answer]
The endpoint doesn't have a Marketplace parameter. [list_financial_events](https://github.com/amzn/selling-partner-api-docs/blob/main/references/finances-api/financesV0.md#get-financesv0financialevents)

I don't understand why you're using a range to loop. Also, if you want to request by NextToken, you only pass NextToken usually, nothing else. You might find the `load_all_pages` and `retry` decorators useful, the do what you want: https://sp-api-docs.saleweaver.com/utils/load_all_pages/

````
@throttle_retry()
@load_all_pages()
def get_financial_events(**kwargs):
    return Finances().list_financial_events(**kwargs)

for page in get_financial_events(PostedAfter='2021-05-10', PostedBefore='2021-05-11', MaxResultsPerPage=100):
    for event in page.payload.get('FinancialEvents').get('ShipmentEventList'):
        print(event)


````

**Comment by paulabirocchi** [comment]
Thank you @saleweaver. This was really helpful.
I'm getting a new error now. 
I'm using the same approach to get historical data, using the period of three months:

```
DATE = '2020-05-01'  # YYYY-MM-DD 
AFTER_DATE='2020-08-01' # YYYY-MM-DD 

@throttle_retry()
@load_all_pages()
def get_financial_events(**kwargs):
    return Finances(marketplace= Marketplaces.NL,credentials=credentials).list_financial_events(**kwargs)
                                     # YYYY-MM-DD           # YYYY-MM-DD

for page in get_financial_events(PostedAfter=DATE, PostedBefore=AFTER_DATE, MaxResultsPerPage=100):

	findict = page.payload
	finnorm = pd.io.json.json_normalize(findict)
```

There is more code after that, I extract all the data from the Finances API. 

The script run until certain time and I've got this error: SellingApiBadRequestException: [{'code': 'InvalidInput', 'message': 'Token is not valid, token duration is 3 minutes.', 'details': ''}]

Can you help me?

**Reply by saleweaver** [reply]
Seems like you're doing a lot with the data after if that takes longer than 3 minutes to complete. I'd push each page to a queue or background process, and request all data within the timeframe amazon allows.

**Reply by paulabirocchi** [reply]
What do you mean with "push each page to a queue or background process" ?

I solved it by creating a loop to get the data for each day from a pandas date_range:

```
`inicio= '2020-01-01'

fim= '2020-05-01'

datas = pd.date_range(start=inicio,end=fim)

for d in np.arange(0,len(datas)-1):

	DATE = str(datas[d])[0:10]  # YYYY-MM-DD 

	AFTER_DATE=str(datas[d+1])[0:10] # YYYY-MM-DD 

	@throttle_retry()
	@load_all_pages()
	def get_financial_events(**kwargs):
	    return Finances(marketplace= Marketplaces.NL,credentials=credentials).list_financial_events(**kwargs)
	                                     # YYYY-MM-DD           # YYYY-MM-DD

	# FOR SPECIFIC DATE:                                       
	for page in get_financial_events(PostedAfter=DATE, PostedBefore=AFTER_DATE, MaxResultsPerPage=100):
		findict = page.payload
		finnorm = pd.io.json.json_normalize(findict)  
```

**Reply by saleweaver** [reply]
Can't you request the entire range? I think 4 months should work.

How I'd do it: 

```

@throttle_retry()
@load_all_pages()
def get_financial_events(**kwargs):
    return Finances(marketplace=Marketplaces.NL, credentials=credentials).list_financial_events(**kwargs)
    # YYYY-MM-DD           # YYYY-MM-DD


# FOR SPECIFIC DATE:                                       
for page in get_financial_events(PostedAfter=DATE, PostedBefore=AFTER_DATE, MaxResultsPerPage=100):
    handle_data_in_background(page.payload)
```

handle_data_in_background spawns a process, or pushes the data onto a queue so that another process can then handle and, for example, save the data. The code in this file would only be responsible for retrieving the data.
